src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,473 @@
+            GetApexTimeDataRoot() + "/etc/tz/tzdata", // Android 10+, TimeData module where the updates land
+            GetApexRuntimeRoot() + "/etc/tz/tzdata",  // Android 10+, Fallback location if the above isn't found or corrupted
+            Environment.GetEnvironmentVariable("ANDROID_DATA") + "/misc/zoneinfo/tzdata",
+            Environment.GetEnvironmentVariable("ANDROID_ROOT") + "/usr/share/zoneinfo/tzdata",
+        };
+
+        private string tzdataPath;
+        private Stream? data;
+
+        private string[]? ids;
+        private int[]? byteOffsets;
+        private int[]? lengths;
+
+        public AndroidTzData(params string[] paths)
+        {
+            foreach (var path in paths)
+            {
+                if (LoadData(path))
+                {
+                    tzdataPath = path;
+                    return;
+                }
+            }
+
+            tzdataPath = "/";
+            ids = new[]{ "GMT" };
+        }
+
+        private static string GetApexTimeDataRoot()
+        {
+            string? ret = Environment.GetEnvironmentVariable("ANDROID_TZDATA_ROOT");
+            if (!string.IsNullOrEmpty(ret!)) {
+                return ret!;
+            }
+
+            return "/apex/com.android.tzdata";
+        }
+
+        private static string GetApexRuntimeRoot()
+        {
+            string? ret = Environment.GetEnvironmentVariable("ANDROID_RUNTIME_ROOT");
+            if (!string.IsNullOrEmpty(ret!))
+            {
+                return ret!;
+            }
+
+            return "/apex/com.android.runtime";
+        }
+
+        private bool LoadData(string path)
+        {
+            if (!File.Exists(path))
+                return false;
+
+            try
+            {
+                data = File.OpenRead(path);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                return false;
+            }
+
+            try
+            {
+                ReadHeader();
+                return true;
+            }
+            catch
+            {
+                // log something here instead of the console.
+                //Console.Error.WriteLine ("tzdata file \"{0}\" was present but invalid: {1}", path, e);
+            }
+            return false;
+        }
+
+        private unsafe void ReadHeader()
+        {
+            int size   = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+            var buffer = new byte[size];
+            var header = ReadAt<AndroidTzDataHeader>(0, buffer);
+
+            header.indexOffset = NetworkToHostOrder(header.indexOffset);
+            header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+            sbyte* s = (sbyte*)header.signature;
+            string magic = new string(s, 0, 6, Encoding.ASCII);
```

Comment:
Why not simply compare the 6 bytes?

Reply:
This one was also directly taken from the mono/mono implementation, I can change it.

Reply:
I'm not exactly sure what criteria should be used to judge incorrectness/corruption of a tzdata file based on its header. I'm having a hard time finding documentation on the tzdata intended file structure.

Reply:
It looks like the beginning of the tzdata file should be "tzdata". The 2012 tzdata file had a header of `"tzdata2012f\0"`, hence the incorrectness if `signature[11] != 0`. The code somewhat follows the check in https://github.com/aosp-mirror/platform_bionic/blob/2ef1cd3f4430d344d88f1275b850119dc963cfa1/libc/tzcode/bionic.cpp#L131-L132
Instead, because a string is being built if the header/signature is incorrect to throw an exception message, I moved the string building out to replace `string magic` building.